File: 07_Studentske_Konsultacije/main.c
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>

#define STUDENT_COUNT 10
#define WAITING_ROOM_SIZE 3

/* FIFO Niz za pracenje ID-eva studenata u cekaonici */
static int waiting_room[WAITING_ROOM_SIZE];
static int head = 0;
static int tail = 0;
static int count = 0;

/* Sinhronizacija rada */
static pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
static pthread_cond_t cond_assistant = PTHREAD_COND_INITIALIZER;
static pthread_cond_t cond_students[STUDENT_COUNT + 1];

static int current_student_in_office = 0;
static int assistant_called = 0;
static int session_finished = 0;
static int remaining_students = STUDENT_COUNT;

static void random_sleep_ms(int min_ms, int max_ms) {
    int delay = min_ms + rand() % (max_ms - min_ms + 1);
    usleep(delay * 1000);
}

static void *assistant_thread(void *arg) {
    (void)arg;

    pthread_mutex_lock(&mutex);
    while (remaining_students > 0) {
        /* Asistent spava ako nema nikoga ispred kabineta */
        while (count == 0 && remaining_students > 0) {
            pthread_cond_wait(&cond_assistant, &mutex);
        }

        if (remaining_students <= 0 && count == 0) break;

        /* FIFO izbacivanje studenta iz reda */
        int student_id = waiting_room[head];
        head = (head + 1) % WAITING_ROOM_SIZE;
        count--;

        current_student_in_office = student_id;
        assistant_called = 1;
        session_finished = 0;

        printf("Asistent proziva studenta %d.\n", student_id);
        fflush(stdout);

        /* Budi se tacno odredjeni student */
        pthread_cond_signal(&cond_students[student_id]);

        /* Asistent ceka da student udje i da obavi konsultacije */
        while (!session_finished) {
            pthread_mutex_unlock(&mutex);
            random_sleep_ms(60, 120); /* Konsultacije */
            pthread_mutex_lock(&mutex);

            if (assistant_called == 0) {
                session_finished = 1;
                printf("Asistent zavrsava konsultacije sa studentom %d.\n", student_id);
                fflush(stdout);
                pthread_cond_signal(&cond_students[student_id]);
            }
        }
    }
    pthread_mutex_unlock(&mutex);
    return NULL;
}

static void *student_thread(void *arg) {
    int id = *(int *)arg;
    free(arg);

    random_sleep_ms(10, 200);

    pthread_mutex_lock(&mutex);
    printf("Student %d dolazi na konsultacije.\n", id);
    fflush(stdout);

    if (count < WAITING_ROOM_SIZE) {
        /* Ima mesta, student seda u cekaonicu */
        waiting_room[tail] = id;
        tail = (tail + 1) % WAITING_ROOM_SIZE;
        count++;

        printf("Student %d seda u cekaonicu.\n", id);
        fflush(stdout);

        pthread_cond_signal(&cond_assistant);

        /* Ceka dok ga asistent ne prozove po imenu */
        while (current_student_in_office != id || !assistant_called) {
            pthread_cond_wait(&cond_students[id], &mutex);
        }

        printf("Student %d ulazi na konsultacije.\n", id);
        fflush(stdout);

        assistant_called = 0; /* Javlja asistentu da je seo */

        /* Ceka signal da su konsultacije gotove */
        pthread_cond_wait(&cond_students[id], &mutex);
        remaining_students--;
    } else {
        printf("Nema mesta u cekaonici. Student %d odlazi.\n", id);
        fflush(stdout);
        remaining_students--;
        pthread_cond_signal(&cond_assistant);
    }

    pthread_mutex_unlock(&mutex);
    return NULL;
}

int main(void) {
    srand((unsigned)time(NULL));
    pthread_t assistant;
    pthread_t students[STUDENT_COUNT];

    for (int i = 1; i <= STUDENT_COUNT; i++) {
        pthread_cond_init(&cond_students[i], NULL);
    }

    pthread_create(&assistant, NULL, assistant_thread, NULL);

    for (int i = 0; i < STUDENT_COUNT; i++) {
        int *id = malloc(sizeof(int));
        *id = i + 1;
        pthread_create(&students[i], NULL, student_thread, id);
    }

    for (int i = 0; i < STUDENT_COUNT; i++) {
        pthread_join(students[i], NULL);
    }

    pthread_mutex_lock(&mutex);
    pthread_cond_signal(&cond_assistant);
    pthread_mutex_unlock(&mutex);

    pthread_join(assistant, NULL);

    pthread_mutex_destroy(&mutex);
    pthread_cond_destroy(&cond_assistant);
    for (int i = 1; i <= STUDENT_COUNT; i++) {
        pthread_cond_destroy(&cond_students[i]);
    }
    printf("\nKonsultacije su zavrsene. Asistent zakljucava kabinet.\n");
    return 0;
}

File: 07_Studentske_Konsultacije/main.go
/*
 * Studentske Konsultacije - Sleeping Barber varijanta sa FIFO
 *
 * -----------------------------------------------------------
 * C vs Go:
 *   pthread_cond_t cond_students[N]  ->  []sync.Cond  (slice kondicionih promenljivih)
 *   pthread_cond_init(&c, NULL)      ->  sync.NewCond(&mu)
 * -----------------------------------------------------------
 */

package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

const (
	STUDENT_COUNT     = 10
	WAITING_ROOM_SIZE = 3
)

var (
	waitingRoom            [WAITING_ROOM_SIZE]int
	head, tail, count      int
	currentStudentInOffice int
	assistantCalled        bool
	sessionFinished        bool
	remainingStudents      int

	mu            sync.Mutex
	condAssistant = sync.NewCond(&mu)
	condStudents  [STUDENT_COUNT + 1]*sync.Cond
)

func init() {
	remainingStudents = STUDENT_COUNT
	for i := 1; i <= STUDENT_COUNT; i++ {
		condStudents[i] = sync.NewCond(&mu)
	}
}

func randomSleep(min, max int) {
	time.Sleep(time.Duration(min+rand.Intn(max-min+1)) * time.Millisecond)
}

func assistantThread(wg *sync.WaitGroup) {
	defer wg.Done()

	mu.Lock()
	for remainingStudents > 0 {
		for count == 0 && remainingStudents > 0 {
			condAssistant.Wait()
		}

		if remainingStudents <= 0 && count == 0 {
			break
		}

		studentID := waitingRoom[head]
		head = (head + 1) % WAITING_ROOM_SIZE
		count--

		currentStudentInOffice = studentID
		assistantCalled = true
		sessionFinished = false

		fmt.Printf("Asistent proziva studenta %d.\n", studentID)
		condStudents[studentID].Signal()

		for !sessionFinished {
			mu.Unlock()
			randomSleep(60, 120)
			mu.Lock()

			if !assistantCalled {
				sessionFinished = true
				fmt.Printf("Asistent zavrsava konsultacije sa studentom %d.\n", studentID)
				condStudents[studentID].Signal()
			}
		}
	}
	mu.Unlock()
}

func studentThread(id int, wg *sync.WaitGroup) {
	defer wg.Done()

	randomSleep(10, 200)

	mu.Lock()
	fmt.Printf("Student %d dolazi na konsultacije.\n", id)

	if count < WAITING_ROOM_SIZE {
		waitingRoom[tail] = id
		tail = (tail + 1) % WAITING_ROOM_SIZE
		count++

		fmt.Printf("Student %d seda u cekaonicu.\n", id)
		condAssistant.Signal()

		for currentStudentInOffice != id || !assistantCalled {
			condStudents[id].Wait()
		}

		fmt.Printf("Student %d ulazi na konsultacije.\n", id)
		assistantCalled = false

		condStudents[id].Wait()
		remainingStudents--
	} else {
		fmt.Printf("Nema mesta u cekaonici. Student %d odlazi.\n", id)
		remainingStudents--
		condAssistant.Signal()
	}

	mu.Unlock()
}

func main() {
	var wg sync.WaitGroup

	wg.Add(1)
	go assistantThread(&wg)

	for i := 1; i <= STUDENT_COUNT; i++ {
		wg.Add(1)
		go studentThread(i, &wg)
	}

	wg.Wait()
	fmt.Println("\nKonsultacije su zavrsene. Asistent zakljucava kabinet.")
}
